Spigot-Server-Patches/0619-added-PlayerTradeEvent.patch

```diff
@@ -4,72 +4,127 @@ Date: Thu, 2 Jul 2020 16:12:10 -0700
+ 
+     @Override
+diff --git a/src/main/java/net/minecraft/world/inventory/Slot.java b/src/main/java/net/minecraft/world/inventory/Slot.java
+index 0b0fa233b817a943ea993260c2c475ef8d669775..88635888d226e13005ee90ed78db0e9abdb614e5 100644
+--- a/src/main/java/net/minecraft/world/inventory/Slot.java
++++ b/src/main/java/net/minecraft/world/inventory/Slot.java
+@@ -32,6 +32,7 @@ public class Slot {
+ 
+     protected void b(int i) {}
+ 
++    protected void checkAdvancements(ItemStack itemStack) { c(itemStack); } // Paper - OBFHELPER
+     protected void c(ItemStack itemstack) {}
+ 
+     public ItemStack a(EntityHuman entityhuman, ItemStack itemstack) {
+diff --git a/src/main/java/net/minecraft/world/inventory/SlotMerchantResult.java b/src/main/java/net/minecraft/world/inventory/SlotMerchantResult.java
+index 8eeb24159d78e3e4b4eac20251a347fa88661892..06e357a3168b61313169ab68a42aa207732ff02a 100644
+--- a/src/main/java/net/minecraft/world/inventory/SlotMerchantResult.java
++++ b/src/main/java/net/minecraft/world/inventory/SlotMerchantResult.java
+@@ -9,9 +9,9 @@ import net.minecraft.world.item.trading.MerchantRecipe;
+ public class SlotMerchantResult extends Slot {
+ 
+     private final InventoryMerchant a;
+-    private final EntityHuman b;
++    private final EntityHuman b; public final EntityHuman getEntityHuman() { return b; } // Paper - OBFEHLPER
+     private int g;
+-    private final IMerchant h;
++    private final IMerchant h; public final IMerchant getMerchant() { return h; } // Paper - OBFHELPER
+ 
+     public SlotMerchantResult(EntityHuman entityhuman, IMerchant imerchant, InventoryMerchant inventorymerchant, int i, int j, int k) {
+         super(inventorymerchant, i, j, k);
+@@ -48,15 +48,29 @@ public class SlotMerchantResult extends Slot {
  
      @Override
-     public boolean isRegularVillager() {
+     public ItemStack a(EntityHuman entityhuman, ItemStack itemstack) {
+-        this.c(itemstack);
+         MerchantRecipe merchantrecipe = this.a.getRecipe();
+ 
++        // Paper start
++        io.papermc.paper.event.player.PlayerTradeEvent event = null;
++        if (getMerchant() instanceof net.minecraft.world.entity.npc.EntityVillagerAbstract) {
++            event = new io.papermc.paper.event.player.PlayerTradeEvent((org.bukkit.entity.Player) getMerchant().getTrader().getBukkitEntity(), (org.bukkit.entity.AbstractVillager) ((net.minecraft.world.entity.npc.EntityVillagerAbstract) getMerchant()).getBukkitEntity(), merchantrecipe.asBukkit(), true, true);
++        }
++        if (event != null) {
++            if (!event.callEvent()) {
++                itemstack.setCount(0);
++                event.getPlayer().updateInventory();
++                return itemstack;
++            }
++            merchantrecipe = org.bukkit.craftbukkit.inventory.CraftMerchantRecipe.fromBukkit(event.getTrade()).toMinecraft();
++            this.checkAdvancements(itemstack);
++        }
```

Comment:
Why are these two blocks?

Reply:
Because as soon as #5583 is merged, there is gonna be an `else if` right after that to construct an event that extends PlayerTradeEvent. so yeah, right now it looks a little strange, but it'll make sense after that other pr is merged.